<a href="https://colab.research.google.com/github/mengcius/pytorch-learn/blob/master/1_%E5%BC%A0%E9%87%8F%E6%95%B0%E6%8D%AE%E7%B1%BB%E5%9E%8B.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 1.PyTorch张量数据类型

In [0]:
import torch
import numpy as np
print(torch.__version__)
print('gpu:', torch.cuda.is_available())

1.1.0
gpu: True


### 数据类型推断

In [0]:
a = torch.randn(2,3) # randn是随机正态分布初始化，(2,3)两维的矩阵
print(a)

print(a.type()) # 打印类型，返回string
print(isinstance(a,torch.FloatTensor)) # 类型判断，合法化检验
type(a) # torch自带的，返回基本数据类型，不常用

tensor([[-0.4501, -0.7879,  0.1924],
        [ 0.0773,  0.4953,  0.0527]])
torch.FloatTensor
True


torch.Tensor

### 转为GPU类型
GPU类型与CPU类型不一样，数值一样但放的位置不一样，则类型不一样

In [0]:
print(isinstance(a,torch.cuda.FloatTensor))

a=a.cuda() # 法1：转为GPU类型，返回一个gpu上的引用
print(isinstance(a,torch.cuda.FloatTensor))

device = torch.device('cuda')
a = a.to(device) # 法2：转为GPU类型
print(isinstance(a,torch.cuda.FloatTensor))

True
True
True


### 形状和维数推断
Dim维数：指Size的长度，看有行列吗，如[*,#]就是2，标量是0

Size/shape：指Tensor形状，看有几行几列，如[2,3]，torch.Size([2, 3])

Tensor：指整个具体数值，如tensor([1.1000, 2.2000])

In [0]:
b=torch.tensor(2.2) # 0维
b.shape # 形状，shape，size

torch.Size([])

In [0]:
len(b.shape) # 维数，shape是成员

0

In [0]:
a=torch.randn(2,3)
a.size() # 形状，size()是成员函数，有括号调用它

torch.Size([2, 3])

In [0]:
a.shape 

torch.Size([2, 3])

In [0]:
len(a.shape) # 维数

2

### 标量（0 维）
Dimension 0 / rank 0，直接输入数值。没有[ ]

In [0]:
a=torch.tensor(1) # 不加点.就是int64的long型，加.就是float32型
print(a)
print(a.type())

tensor(1)
torch.LongTensor


In [0]:
torch.tensor(1.3)

tensor(1.3000)

### 1 维
Dim 1 / rank 1，数学里叫向量，pytorch里不管几维都是Tensor。引入维数要加[ ]。

1维的数一般用在偏置bias；神经网络线性层需要将一张图片展平来作为输入；pytorch0.3版的loss返回为1维[]，而新版就改为了标量。

.tensor()接受的是数据的内容；.FloatTensor()接受的是数据的shape维数，初始化就用randn。

In [0]:
torch.tensor([1.1]) # 长度为1的向量也是向量，虽然它也可以表达标量，但为了概念混乱还是要和标量区分开

tensor([1.1000])

In [0]:
torch.tensor([1.1,2.2]) # 长度为2的向量

tensor([1.1000, 2.2000])

In [0]:
torch.FloatTensor(2) # 长度为2的向量

tensor([4.1556e+12, 0.0000e+00])

从numpy引入数据

In [0]:
data=np.ones(2) # 长度为2的向量
print(data)
data

[1. 1.]


array([1., 1.])

numpy转为torch

In [0]:
torch.from_numpy(data) # numpy转为torch

tensor([1., 1.], dtype=torch.float64)

In [0]:
data.shape

(2,)

In [0]:
len(data.shape)

1


### 2 维
如网络的输入图像矩阵[batch,718]

In [0]:
a=torch.randn(2,3) # 随机正态分布
a

tensor([[-0.5858, -0.8323, -0.6999],
        [ 2.1577,  1.4279,  0.1812]])

In [0]:
a.size()

torch.Size([2, 3])

In [0]:
a.shape

torch.Size([2, 3])

In [0]:
print(a.size(0))
print(a.size(1))
print(a.shape[0])
print(a.shape[1])

2
3
2
3


### 3 维
适合RNN的文字输入 [10,20,100]：10个单词、20个句子、每单词100个特征

In [0]:
a=torch.randn(1,2,3) # 随机均匀分布，0-1间
a # 第一个维度为1，所以最外面的中括号里有一个元素

tensor([[[-0.2602, -0.9290, -1.7684],
         [ 1.5547, -1.6179, -1.0734]]])

In [0]:
a.size()

torch.Size([1, 2, 3])

In [0]:
a[0]

tensor([[-0.2602, -0.9290, -1.7684],
        [ 1.5547, -1.6179, -1.0734]])

In [0]:
list(a.shape) #把shape返回的类型torch.Size直接转为list

[1, 2, 3]

### 4 维
适合CNN里彩色图片的输入，[batch,c,h,w]

In [0]:
a=torch.randn(1,3,5,5)
a

tensor([[[[ 1.4470, -0.6711,  0.1644, -0.4992,  0.6585],
          [ 0.5644, -1.0187,  2.1814,  1.2801,  0.5476],
          [-0.6059, -0.9888,  0.1079, -0.6363, -0.4275],
          [-0.3597,  0.1857, -0.1193, -0.2643,  0.0285],
          [ 0.0800,  1.3303, -0.7817, -1.5925,  0.0633]],

         [[ 3.2000, -0.6776, -0.7467, -1.6329, -0.9923],
          [-0.0510, -0.1238,  0.7909,  0.1247, -0.5297],
          [-1.1373, -1.1427,  0.1506, -1.2995, -0.1222],
          [ 0.2495,  1.8472,  0.0956,  0.1645,  0.1028],
          [ 0.7881,  1.0822, -0.4555, -0.9690, -0.8317]],

         [[-0.3532, -0.4538,  1.3515,  1.5554, -0.2043],
          [ 2.3563, -0.0954,  0.3800, -1.4728,  0.1570],
          [ 0.3729, -1.2788,  0.2015,  2.6024, -0.6739],
          [-0.8144,  0.2691, -1.1654, -1.7682,  1.1866],
          [-1.0817,  1.0890,  2.2866, -0.7912, -1.6249]]]])

In [0]:
a.numel() #numel，大小，指tensor占用内存的数量，就是所有元素的数量

75

In [0]:
a.dim() #维数，更直观

4

In [0]:
len(a.shape)

4